### PR TITLE
Couple of changes to make it easier to compile rustc for wasm

### PR DESCRIPTION
**Pull request by bjorn3**
This is a subset of the patches I have on my rust fork to compile rustc for wasm32-wasip1.

### CODE REVIEW

**Comment by bjorn3** [comment]
cc https://github.com/rust-lang/miri/issues/722

**Review by davidtwco — APPROVED** [review]

**Comment by onur-ozkan** [comment]
Apparently this is being problem for cross compilations (https://github.com/rust-lang/rust/issues/135796, https://github.com/rust-lang/rust/issues/133629).

**Comment by bjorn3** [comment]
I'm guessing this didn't fail in CI as we default to using rpaths, while these distros/OSes don't?

File: src/bootstrap/src/core/builder.rs [hunk 1]
```diff
@@ -1683,10 +1683,24 @@ impl<'a> Builder<'a> {
...
+                        .arg("-")
+                        .run_capture(self)
+                        .stderr();
+                    let not_supported = error
+                        .lines()
+                        .any(|line| line.contains("unsupported crate type `proc-macro`"));
+                    if !not_supported {
+                        cargo.arg("-Zdual-proc-macros");
+                        rustflags.arg("-Zdual-proc-macros");
+                    }
```
Comment (onur-ozkan) [inline src/bootstrap/src/core/builder.rs:1703]: Would adding `rpath` check here fix the regression?

Comment (bjorn3) [reply src/bootstrap/src/core/builder.rs:1703]: I think `self.rustc()` may be missing code to set `LD_LIBRARY_PATH`. If `LD_LIBRARY_PATH` is set correctly it should work even with rpath disabled.

**Pull request merged**